Account model

drkitt opened the pull request:
## Tickets

#27

## Summary Of Changes
- Set up view, business, and query layers in the `account` app
- Created User model to store a user's account
- Added functionality to get and create a user
- Added functionality to update a user's information

## Notes for Reviewers
- To run the backend tests, go into the terminal on the backend Docker container (I go to View > Terminal in VS Code with the backend container open, but you can also access it through Docker Desktop). In the terminal, run the command `./manage.py test`.
- The business functions are untested because they're just wrappers for the query function
- A lot of the query logic is temporary, to be replaced once we can get data from the frontend (#29) and the user's Google account (#26)
- I'm not sure if the Google OAuth keys are useful for uniquely accessing the user from the frontend (e.g. the `update_user` query function), especially considering that all Django models automatically get a unique `id` to use as a primary key, so any input on that is appreciated

jaredwebber commented on backend/account/migrations/0001_initial.py:
```diff
@@ -0,0 +1,50 @@
+            name='User',
+            fields=[
+                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('password', models.CharField(max_length=128, verbose_name='password')),
```
Password should be handled entirely by OAuth right? Is this field used anywhere?

jaredwebber replied:
Also not sure where `is_superuser`, or `is_staff`would be used - got any examples?

drkitt replied:
Those are both from Django's `AbstractUser` class and aren't currently in use in this project. I personally think we should keep the `AbstractUser` inheritance because it has a bunch of convenient helper methods, but if we decide we don't want to have those unused fields then we can just reimplement `AbstractUser`'s helper methods as needed in our own `User` class.

jaredwebber replied:
Ah sure if if they're inherited doesn't really matter to me if we keep them 👍

jaredwebber reviewed (approved):
LGTM - dunno if @Speuce wants to take a look, not too confident about model best practices

Speuce commented on backend/account/business.py:
```diff
@@ -0,0 +1,15 @@
+
+
+def update_user(payload: dict):
+    query.update_user(payload)
```
Newline at end of file. Need to check pylint, it should be catching this.
```suggestion
    query.update_user(payload)

```

Speuce commented on backend/account/query.py:
```diff
@@ -0,0 +1,39 @@
+    # Since this is at the end of the loop, it'll only execute if an exception
+    # isn't raised. That is, we'll only save changes if the entire payload is
+    # error-free.
+    user.save()
```
```suggestion
    user.save()

```

Speuce commented on backend/account/models.py:
```diff
@@ -1,3 +1,11 @@
-# Create your models here.
+class User(AbstractUser):
+    # Name, username, and email are included with AbstractUser
+    google_id = models.TextField()
```
Can we enforce a unique constraint on google_id?

Speuce reviewed (changes requested):
Looks good overall just a couple very minor changes

drkitt commented:
Fixed Pylint errors and added the unique constraint. The new constraint doesn't have any tests because it isn't actually possible to violate it with the current operations (`get_or_create` will just get a user if you give it an id that's in use, and `update` uses `google_id` to find a record and thus can't give it a new value). I agree that we should have that constraint, just explaining why there aren't any new tests.

Speuce commented:
Sorry for the close/open just want to force django CI to run here too

Speuce reviewed (approved):
LGTM if actions pass
